api4/system.go

```diff
@@ -461,6 +461,11 @@ func testS3(c *Context, w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	if cfg.FileSettings.IsEmpty() {
```

Comment:
I think the missing part here was to do `cfg.SetDefaults()`. Let's do that immediately after `if cfg == nil` here and in `testEmail` as well.

Reply:
Hmm, makes sense but thinking if it lead to some confusion as these APIs are public. I mean we set the defaults but don't let them know if they provided missing values etc. Because the Local and S3 config is different. Let's think of a scenario they provided everything for S3 but not the driver name, if we set to default, the validation process will go to Local driver validation path. (or processing path).

Reply:
The `SetDefaults` is done in api4.updateConfig as well. That's why I was thinking of making it consistent everywhere.

But you are right that silently validating local driver path instead of s3 path is also not right. Can we do a reflection based approach where we just iterate through all the pointer fields in `FileSettings` and throw an error if anything is nil? My main concern is to not have to add yet another field when we add a new field in that struct.

Apart from this, let's also move the json parsing inside and throw an error if it fails. So these are my suggestions:
1. Do `json.NewDecoder(data).Decode(&o)` check inside and check for error.
2. Check if cfg.FileSettings is nil or not.
3. Add a private function to just iterate through all fields in cfg.FileSettings and check if any is nil.

Does this sound reasonable?

Reply:
Hmm, but with this approach, we still need to update the `reflect` validation because if we use the `ImageDriverLocal` we shouldn't check if S3 related fields are nil or not right? Or, the user must set all the fields to some sort of a value to avoid failing in Local driver or vice versa.

Agreed in error check on decode btw. But `FileSettings` is not a pointer so we still need to check if it's empty one way or another.

The issue here is forcing the flat data structure for a completely different data structures. And it's quite messy to make this work.

Reply:
> we still need to update the reflect validation because if we use the ImageDriverLocal we shouldn't check if S3 related fields are nil or not right? 

I was thinking that we might not need to make it that complicated. The config, by default always fills up some empty fields. So nil pointers are anyways not valid. Instead of filling them up with default values, we are just throwing an error here.

Reply:
Oh, right. So, if someone wants to validate their `Config.FileSettings`, they should at least, do a `cfg.SetDefaults()` beforehand. And you are proposing the reflect based `nil` check to be added here right? https://github.com/mattermost/mattermost-server/blob/82c8bd33ed1172c03c9f46e32972afba88f9c610/model/config.go#L1491  Because the panic occurred here. Or, you prefer it added to FileSettings validation and run every time (which is triggered by `Config.Validate` as it's a private method). Asking this because we actually do not validate the config as a whole for this API.